### ISSUE TITLE
EPIC: Cryptographic verification of equivocation

### ISSUE THREAD

**Issue by jtremback** [cosmos/interchain-security#732]
# Problem

We would like the provider chain to be able to verify equivocation evidence on its own, instead of trusting consumers, or needing evidence to go through governance. For this, we can look at the code in Tendermint which does this, but it will need heavy changes.

# Problem details

To keep things simple at first, we should design this code to be used by an accuser who would like to cause an equivocating validator to be slashed by submitting a message to the chain containing all necessary evidence. Later, we can adapt this to be used automatically for packets coming from a consumer chain.

Fortunately, almost all the code used by Tendermint to verify equivocation evidence is in one file: https://github.com/cometbft/cometbft/blob/main/evidence/verify.go. A lot of this code should be possible to reuse, however, it depends on state stored by Tendermint nodes, most of which will not be available on the provider chain.

There are two basic types of equivocation: Double signing and light client attacks.

## Double signing

This is trivial, you can see the function that does it [here](https://github.com/cometbft/cometbft/blob/main/evidence/verify.go#L160). The only complicated thing is that it requires the validator set and powers at the height of the evidence. But we should be able to store these on the provider. Also, the evidence will have the height of the consumer which will need to be mapped to the height of the provider to get the validator powers at that height. But I think we already have code to do that.

## Light client attacks

This is much [more complicated](https://github.com/cometbft/cometbft/blob/main/evidence/verify.go#L57), and it requires more state. The additional state that is required is the true header of the block in question, known as the "trustedHeader" or the "commonHeader". In Tendermint, this is loaded from the node's state. On the provider chain, we will not be able to access this. Instead, it should be possible to have the accuser supply the trusted header along with the rest of the evidence. The provider chain can then use the record of validator powers from that block to check that the trusted header is real.

# Task list

```[tasklist]
### Must have
- [ ] https://github.com/cosmos/interchain-security/issues/1124
- [ ] https://github.com/cosmos/interchain-security/issues/825
- [ ] https://github.com/cosmos/interchain-security/issues/827
- [ ] https://github.com/cosmos/interchain-security/issues/1147
- [ ] https://github.com/cosmos/interchain-security/issues/1065
- [ ] https://github.com/cosmos/interchain-security/issues/1224
- [ ] https://github.com/cosmos/interchain-security/issues/1227
- [x] Fix consumer public keys retrieval in double voting evidence handling: #1261 #1254
- [x] Wait on the Hermes relayer fix to detect both ICS misbehaviour and double voting handling; blocking #1255
- [ ] https://github.com/cosmos/interchain-security/issues/1277
- [x] Support Hypha with the Gaia v13 tests
- [ ] https://github.com/cosmos/interchain-security/issues/1255
- [ ] https://github.com/cosmos/interchain-security/issues/1266
- [ ] https://github.com/cosmos/interchain-security/issues/1026
- [ ] https://github.com/cosmos/interchain-security/issues/1222
- [ ] https://github.com/cosmos/interchain-security/issues/1274
- [x] Wait on the slashing to be integrated; putting #1302 temporarlily on hold
- [ ] https://github.com/cosmos/interchain-security/issues/1302
- [ ] https://github.com/cosmos/interchain-security/issues/1400
- [ ] https://github.com/cosmos/interchain-security/pull/1294
- [ ] https://github.com/cosmos/interchain-security/issues/1346
- [ ] https://github.com/cosmos/interchain-security/issues/1387
- [ ] https://github.com/cosmos/interchain-security/issues/1403
```

```[tasklist]
### Nice to have
- [ ] https://github.com/cosmos/interchain-security/issues/1125
- [ ] https://github.com/cosmos/interchain-security/issues/1180
```

**Comment by sainoe** [comment]
Summarizing our last discussion with @mpoke, we can separate the work into two main concerns:

1. Define the requirements to migrate the equivocation evidence verification from Tendermint to the ICS module.

2. Figure out what it implies to get an equivocation evidence from a malicous consumer and how we can leverage the information stored by IBC.

**Comment by tbruyelle** [comment]
About 2.

> Also, the evidence will have the height of the consumer which will need to be mapped to the height of the provider to get the validator powers at that height. But I think we already have code to do that.

What we have currently is a mapping that uses the valset update id. On the provider chain, this id maps a block height. If we decide to use it, that means the equivocation evidence message must contain the valset id, so the provider can determine the height, and then determine the voting power at this height.

My concern is how do we verify that valset id passed in the message is correct ? I don't think we can use IBC client for that.

Maybe instead the message should contain the entire valset, instead of just the ID. At least the entire valset can be verified if I'm not wrong. But then the message is harder to create for the user.

**Comment by ancazamfir** [comment]
Or we do not trust that consumer chain validators run the proper tendemint protocols, for example they may drop all evidence or duplicate votes without recording them?

**Comment by cason** [comment]
I would suggest to update the links to the CometBFT [repository](https://github.com/cometbft/cometbft). The code should be the same, but the Tendermint Core (old) repository is now obsolete.
